### PR TITLE
Reverse pagination

### PR DESCRIPTION
**Pull request by aslowwriter**
Sanity check:

* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/getzola/zola/pulls) for the same update/change?
* [X] Are you doing the PR on the `next` branch?
* [X] Have you created/updated the relevant documentation page(s)?

# Description
This PR is in response to issue #776. I've wanted the reverse pagination myself so I've implemented it. I know the wat this should be exposed to the user hasn't been agreed upon yet, but I wanted to make the PR anyway for the following reasons:

1. Since pagination happens before the template get's access to the paginator it was clear to me that this would have to be set in the section front_matter so it could be handled in time. 
2. the functionality is the same regardless of how it's handled in the front matter
3. changing the names of the sort_by options is quite easy. 

Given these observations, I think the option I went with (e.g. `weight_asc` and `weight_desc` instead of `weight`) is the most organic/consistent. However, if people prefer backwards-compatibility, which would also be fair, we can revert the names of the `sort_by` and add a `paginate_reversed` option, this would be a very easy change (that was actually how I did it originally). 
If people want a different naming altogether that is also possible. I wanted to make the PR anyway so we have something to use during the discussions. I'm open to change requests or other feedback 

P.s. I also removed the `order` field in the front matter of the page since this should now be obsolete and wasn't used anywhere else.

### CODE REVIEW

**Comment by Keats** [comment]
> Given these observations, I think the option I went with (e.g. weight_asc and weight_desc instead of weight) is the most organic/consistent. However, if people prefer backwards-compatibility, which would also be fair, we can revert the names of the sort_by and add a paginate_reversed option, this would be a very easy change (that was actually how I did it originally).

I think it's better to go the `paginate_reversed` way, the default order should be fine for most people , I'll review the PR once it's switched

**Comment by aslowwriter** [comment]
@Keats Done

File: components/front_matter/src/lib.rs [hunk 1]
```diff
@@ -19,11 +19,8 @@ lazy_static! {
 #[derive(Debug, Copy, Clone, PartialEq, Serialize, Deserialize)]
 #[serde(rename_all = "lowercase")]
 pub enum SortBy {
-    /// Most recent to oldest
     Date,
-    /// Lower weight comes first
     Weight,
-    /// No sorting
     None,
```
Comment (Keats) [inline components/front_matter/src/lib.rs]: why remove those comments?

File: components/library/src/content/ser.rs [hunk 2]
```diff
@@ -150,6 +150,10 @@ impl<'a> SerializingPage<'a> {
         }
     }
 
+    // just for testing the paginator
```
Comment (Keats) [inline components/library/src/content/ser.rs]: can that be removed?

File: components/library/src/pagination/mod.rs [hunk 3]
```diff
@@ -262,15 +273,22 @@ mod tests {
...
-        library.insert_page(Page::default());
-        library.insert_page(Page::default());
-        library.insert_page(Page::default());
+    fn create_library(
+        is_index: bool,
+        num_pages: usize,
+        paginate_reversed: bool,
+    ) -> (Section, Library) {
+        let mut library = Library::new(num_pages, 0, false);
+        for i in 1..num_pages + 1 {
```
Comment (Keats) [inline components/library/src/pagination/mod.rs]: isn't `...` equivalent to `end + 1`?

File: components/site/src/lib.rs [hunk 4]
```diff
@@ -901,6 +901,24 @@ impl Site {
...
+            self.render_feed(
+                pages,
+                Some(&PathBuf::from(&section.path[1..])),
+                &section.lang,
+                |mut context: Context| {
+                    context.insert("section", &section.to_serialized(library));
+                    context
+                },
+            )?;
+        }
```
Comment (Keats) [inline components/site/src/lib.rs]: looks like there are unrelated changes in the commits

File: components/library/src/pagination/mod.rs [hunk 5]
```diff
@@ -45,11 +45,13 @@ impl<'a> Pager<'a> {
 #[derive(Clone, Debug, PartialEq)]
 pub struct Paginator<'a> {
     /// All pages in the section/taxonomy
-    all_pages: &'a [DefaultKey],
+    all_pages: Vec<DefaultKey>,
```
Comment (Keats) [inline components/library/src/pagination/mod.rs]: Would `Cow` work here to avoid the performance penalty when cloning the keys if we don't need to?

**Review by Keats — CHANGES_REQUESTED** [review]
Can you clean the commits? I think a rebase one `next` should work, or manually removing the unrelated commits from the history

File: components/front_matter/src/lib.rs (hunk 1 above)
Comment (aslowwriter) [reply components/front_matter/src/lib.rs]: A mistake on my part when reverting from the `weight_asc` config. I'll revert it

File: components/library/src/content/ser.rs (hunk 2 above)
Comment (aslowwriter) [reply components/library/src/content/ser.rs]: sure

File: components/library/src/pagination/mod.rs (hunk 3 above)
Comment (aslowwriter) [reply components/library/src/pagination/mod.rs]: I looked it up `..=` is an inclusive range. I'll make that change if you prefer,

File: components/site/src/lib.rs (hunk 4 above)
Comment (aslowwriter) [reply components/site/src/lib.rs]: probably my mistake while pulling `next` so the PR wouldn't have merge conflicts, I'll remove it

File: components/library/src/pagination/mod.rs (hunk 5 above)
Comment (aslowwriter) [reply components/library/src/pagination/mod.rs]: good suggestion, I added it in the next commit

**Comment by aslowwriter** [comment]
My git-fu isn't too great so I'm struggling a little to get the commit history correct and cleaned up but I'm working on it

**Comment by aslowwriter** [comment]
Okay, I think I got rid of all unrelated changes now. I tried to clean up the commit history as well as I could too, I hope it's okay like this

File: CHANGELOG.md [hunk 6]
```diff
@@ -22,6 +22,7 @@ binary size
 - Add Markdown-outputting shortcodes
 - Add `minify_html` option to `config.toml` to minify HTML automatically
 
+
```
Comment (Keats) [inline CHANGELOG.md]: you can add this change to that file ;)

File: docs/content/documentation/content/section.md [hunk 7]
```diff
@@ -164,6 +165,13 @@ to newest (at the bottom).
 
 `reverse` has no effect on `page.later`/`page.earlier` or `page.heavier`/`page.lighter`.
 
+However this will not work correctly if the pages are paginated, because it
+will only apply to the pagers and not to all pages. For example if you have
+pages `[1,2,3,4]` and a `paginate_by=2` they will be paginated as
+`[[2,1],[4,3]]` instead of `[[4,3],[2,1]]` as you might expect. To achive this
```
Comment (Keats) [inline docs/content/documentation/content/section.md]: achieve

File: docs/content/documentation/content/section.md [hunk 8]
```diff
@@ -156,6 +156,7 @@ This will be sort all pages by their `weight` field, from lightest weight
 page gets `page.lighter` and `page.heavier` variables that contain the
 pages with lighter and heavier weights, respectively.
 
+### reversed sorting
```
Comment (Keats) [inline docs/content/documentation/content/section.md]: Reversed sorting

File: docs/content/documentation/getting-started/configuration.md [hunk 9]
```diff
@@ -88,6 +88,10 @@ compile_sass = false
 # When set to "true", the HTML output files will be minified
 minify_html = false
 
+# When set to "true", a search index is built from the pages and section
+# content for `default_language`.
+build_search_index = false
```
Comment (Keats) [inline docs/content/documentation/getting-started/configuration.md]: Isn't it already in there?

File: components/library/src/content/ser.rs [hunk 10]
```diff
@@ -150,6 +150,10 @@ impl<'a> SerializingPage<'a> {
         }
     }
 
+    pub fn get_title(&'a self) -> &'a Option<String> {
```
Comment (Keats) [inline components/library/src/content/ser.rs:154]: Can you add back the docstring saying it's only used in tests?

**Review by Keats — CHANGES_REQUESTED** [review]
Looks like we need to add `.swp` to gitignore

File: docs/content/documentation/content/section.md (hunk 7 above)
Comment (aslowwriter) [reply docs/content/documentation/content/section.md]: does that mean you want this part removed?

Comment (Keats) [reply docs/content/documentation/content/section.md]: no, just the typo

**Review by Keats — APPROVED** [review]

**Comment by Keats** [comment]
Thanks!
